mrjob/dataproc.py

```diff
@@ -0,0 +1,873 @@
+        step with the given ID to complete, and fetch counters.
+        If it fails, attempt to diagnose the error, and raise an
+        exception.
+
+        This also adds an item to self._log_interpretations
+        """
+        log_interpretation = dict(job_id=job_id)
+        self._log_interpretations.append(log_interpretation)
+
+        while True:
+            # don't antagonize Dataproc's throttling
+            log.debug('Waiting %.1f seconds...' %
+                      self._opts['check_api_status_every'])
+            time.sleep(self._opts['check_api_status_every'])
+
+            # https://cloud.google.com/dataproc/reference/rest/v1/projects.regions.jobs#JobStatus
+            job_result = self._get_job(job_id)
+
+            job_start_time = job_result['status']['stateStartTime']
+            job_state = job_result['status']['state']
+
+            log.info('%s  %s - %s' % (job_id, job_start_time, job_state))
+
+            # TODO - mtai @ davidmarin - State checking is NOT tested at this point, see included URL for Dataproc Job states
+            # NOTE -  mtai @ davidmarin - https://cloud.google.com/dataproc/reference/rest/v1/projects.regions.jobs#State
+            if job_state not in ('DONE', 'ERROR', 'CANCELLED'):
+                continue
+
+            # we're done, will return at the end of this
+            if job_state == 'DONE':
+                break
+
+            elif job_state == 'ERROR':
+                raise StepFailedException(step_num=step_num, num_steps=num_steps)
+
+    def _step_input_uris(self, step_num):
+        """Get the gs:// URIs for input for the given step."""
+        if step_num == 0:
+            return [self._upload_mgr.uri(path)
+                    for path in self._get_input_paths()]
+        else:
+            # put intermediate data in HDFS
+            return ['hdfs:///tmp/mrjob/%s/step-output/%s/' % (
+                self._job_key, step_num)]
+
+    def _step_output_uri(self, step_num):
+        if step_num == len(self._get_steps()) - 1:
+            return self._output_dir
+        else:
+            # put intermediate data in HDFS
+            return 'hdfs:///tmp/mrjob/%s/step-output/%s/' % (
+                self._job_key, step_num + 1)
+
+    def counters(self):
+        return [_pick_counters(log_interpretation)
+                for log_interpretation in self._log_interpretations]
+
+    ### Bootstrapping ###
+
+    def _create_master_bootstrap_script_if_needed(self):
+        """Helper for :py:meth:`_add_bootstrap_files_for_upload`.
+
+        Create the master bootstrap script and write it into our local
+        temp directory. Set self._master_bootstrap_script_path.
+
+        This will do nothing if there are no bootstrap scripts or commands,
+        or if it has already been called."""
+        if self._master_bootstrap_script_path:
+            return
+
+        # TODO - mtai @ davidmarin - Don't know when we don't need to do this...
```

Comment:
Definitely need to do this; offhand I can't think of any way that Dataproc differs from EMR here.